include/OGL/CommunicationPattern.H

```diff
@@ -0,0 +1,148 @@
+ * @param size how many elements to send to owner
+ * @param total_size Size of the total buffer
+ * @param padding_before Number of unused elements before
+ * @param padding_after Number of unused elements after
+ * @return The AllToAllPattern
+ * */
+AllToAllPattern compute_gather_to_owner_counts(
+    const ExecutorHandler &exec_handler, label ranks_per_owner, label size,
+    label total_size, label padding_before, label padding_after);
+
+/* @brief  This function computes the send and recv counts and offset vectors
+ *for repartitioning to an owner rank
+ *
+ * @param exec_handler The executor handler
+ * @param ranks_per_owner Ratio of total ranks to owner ranks
+ * @parameter size how many elements to send to owner
+ * @return The AllToAllPattern
+ * */
+AllToAllPattern compute_gather_to_owner_counts(
+    const ExecutorHandler &exec_handler, label ranks_per_owner, label size);
+
+/** @brief Given a rank id and ranks_per_owner this function computes the
+ * corresponding owner rank
+ *
+ * @param rank The rank under consideration
+ * @param ranks_per_owner Ratio of total ranks to owner ranks
+ * @return The corresponding owner of this rank
+ * */
+label compute_owner_rank(label rank, label ranks_per_owner);
+
+/* @brief Given a comm_pattern this function executes the communication
+ *
+ * The purpose of this function is to have a uniform function for communication
+ * calls so that we can inspect the communication for debugging purposes
+ *
+ * @param exec_handler The executor handler
+ * @param comm_pattern The communication pattern i.e. the send, recv counts and
+ * offsets
+ * @param send_buffer Pointer to the data buffer to send
+ * @param recv_buffer Pointer to the data buffer to recv
+ */
+void communicate_values(const ExecutorHandler &exec_handler,
+                        const AllToAllPattern &comm_pattern,
+                        const scalar *send_buffer, scalar *recv_buffer);
+
+/* @brief A function that can perform an offseted gather to owner rank
+ *
+ * This function is similar to communicate_values. However, it takes care of
+ * allocating the revc_buffer on the owner rank and can add a prescribed offset
+ * to all elements. The latter is often required when gathering indices to the
+ * owner rank.
+ *
+ * @param exec_handler The executor handler
+ * @param send_buffer Pointer to send data
+ * @param send_size How many elements to send to owner
+ * @param offset Value to add to all elements in the send_buffer
+ * Returns a vector of labels with a length equal the total received elements +
+ * padding */
+std::vector<label> gather_labels_to_owner(const ExecutorHandler &exec_handler,
+                                          const AllToAllPattern &comm_pattern,
+                                          const label *send_buffer,
+                                          label send_size, label offset = 0);
+
+struct CommunicationPattern {
+    using comm_size_type = label;
```

Comment:
`comm_size_type` is defined as a type alias of the type `label` here, but sometimes the type alias `comm_size_type` and the type `label` are used together in the following declarations, e.g., the declaration of the variable `send_idxs `. 
It causes some confusion.

Reply:
Agree, i guess we can use label `everywhere`